### PR TITLE
MAILBOX-357 Only use Mailbox APIs for MailboxEvents

### PR DESCRIPTION
**Pull request by chibenwa**
Store events where used for two features:
 - Avoiding reading the mailbox from the store
 - Caching messages to save some reads

We need to rely on Mailbox API and read mailbox/messages from the MailboxFactory.

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java [hunk 1]
```diff
@@ -66,13 +66,11 @@ public void event(Event event) {
...
+                    Mailbox mailbox = added.getMailbox();
 
-                    for (final MessageUid next : (Iterable<MessageUid>) added.getUids()) {
-                        Optional<MailboxMessage> mailboxMessage = retrieveMailboxMessage(session, added, mailbox, next);
-                        if (mailboxMessage.isPresent()) {
-                            addMessage(session, mailbox, mailboxMessage.get());
-                        }
+                    for (MessageUid next : added.getUids()) {
+                        retrieveMailboxMessage(session, added, mailbox, next)
+                            .ifPresent(mailboxMessage1 -> addMessage(session, mailbox, mailboxMessage1));
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java]: s/mailboxMessage1/mailboxMessage

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java [hunk 2]
```diff
@@ -97,19 +97,14 @@ public void event(Event event) {
...
-                LOGGER.error("Could not retrieve message {} in mailbox {}", next, mailbox.getMailboxId().serialize(), e);
-                return Optional.empty();
-            }
+    private Optional<MailboxMessage> retrieveMailboxMessage(MailboxSession session, Mailbox mailbox, MessageUid next) {
+        try {
+            return Optional.of(factory.getMessageMapper(session)
+                .findInMailbox(mailbox, MessageRange.one(next), FetchType.Full, UNLIMITED)
+                .next());
+        } catch (Exception e) {
+            LOGGER.error("Could not retrieve message {} in mailbox {}", next, mailbox.getMailboxId().serialize(), e);
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java]: s/next/uid + implicit toString call

File: mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassinListener.java [hunk 3]
```diff
@@ -68,32 +81,50 @@ public ExecutionMode getExecutionMode() {
 
     @Override
     public void event(Event event) {
```
Comment (chibenwa) [inline mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassinListener.java]: Method extraction ^^ it is too big

File: mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassinListener.java [hunk 4]
```diff
@@ -68,32 +81,50 @@ public ExecutionMode getExecutionMode() {
...
-                List<InputStream> contents = addedEvent.getAvailableMessages()
-                    .values()
-                    .stream()
-                    .map(Throwing.function(MailboxMessage::getFullContent))
-                    .collect(Guavate.toImmutableList());
-                spamAssassin.learnHam(contents, event.getUser());
-            }
+    }
+
+    private Iterator<MailboxMessage> retrieveMessage(MessageMapper messageMapper, Mailbox mailbox, MessageUid uid) {
```
Comment (chibenwa) [inline mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassinListener.java]: Return type as a Stream

File: mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassinListener.java [hunk 5]
```diff
@@ -68,32 +81,50 @@ public ExecutionMode getExecutionMode() {
...
+        try {
+            return messageMapper.findInMailbox(mailbox, MessageRange.one(uid), MessageMapper.FetchType.Full, LIMIT);
+        } catch (MailboxException e) {
+            LOGGER.warn("Can not retrieve message {} {}", mailbox.getMailboxId(), uid, e);
```
Comment (chibenwa) [inline mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassinListener.java]: Implicit toString call

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java [hunk 6]
```diff
@@ -63,34 +62,35 @@ public ListeningMessageSearchIndex(MailboxSessionMapperFactory factory, MailboxM
     public void event(Event event) {
         try {
             MailboxSession session = mailboxManager.createSystemSession(event.getUser().asString());
-            if (event instanceof MessageEvent) {
-                if (event instanceof EventFactory.AddedImpl) {
-                    EventFactory.AddedImpl added = (EventFactory.AddedImpl) event;
-                    Mailbox mailbox = added.getMailbox();
+            if (event instanceof MailboxEvent) {
+                MailboxEvent messageEvent = (MailboxEvent) event;
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java]: s/messageEvent/mailboxEvent

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java [hunk 7]
```diff
@@ -97,13 +99,13 @@ public void event(Event event) {
         }
     }
 
-    private Optional<MailboxMessage> retrieveMailboxMessage(MailboxSession session, Mailbox mailbox, MessageUid next) {
+    private Optional<MailboxMessage> retrieveMailboxMessage(MailboxSession session, Mailbox mailbox, MessageUid uid) {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java]: Return a stream...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java [hunk 8]
```diff
@@ -97,13 +99,13 @@ public void event(Event event) {
...
-    private Optional<MailboxMessage> retrieveMailboxMessage(MailboxSession session, Mailbox mailbox, MessageUid next) {
+    private Optional<MailboxMessage> retrieveMailboxMessage(MailboxSession session, Mailbox mailbox, MessageUid uid) {
         try {
             return Optional.of(factory.getMessageMapper(session)
-                .findInMailbox(mailbox, MessageRange.one(next), FetchType.Full, UNLIMITED)
+                .findInMailbox(mailbox, MessageRange.one(uid), FetchType.Full, UNLIMITED)
                 .next());
         } catch (Exception e) {
-            LOGGER.error("Could not retrieve message {} in mailbox {}", next, mailbox.getMailboxId().serialize(), e);
+            LOGGER.error("Could not retrieve message {} in mailbox {}", uid, mailbox.getMailboxId().serialize(), e);
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java]: Implicit toString

**Comment by chibenwa** [comment]
```
[e296c51ebb3b1b6c316497c7573008e6e29e3f30] [ERROR]   SpamAssassinListenerTest.eventShouldCallSpamAssassinHamLearningWhenTheMessageIsAddedInInbox:259 
[e296c51ebb3b1b6c316497c7573008e6e29e3f30] Wanted but not invoked:
[e296c51ebb3b1b6c316497c7573008e6e29e3f30] spamAssassin.learnHam(<any>, <any>);
[e296c51ebb3b1b6c316497c7573008e6e29e3f30] -> at org.apache.james.mailbox.spamassassin.SpamAssassinListenerTest.eventShouldCallSpamAssassinHamLearningWhenTheMessageIsAddedInInbox(SpamAssassinListenerTest.java:259)
[e296c51ebb3b1b6c316497c7573008e6e29e3f30] Actually, there were zero interactions with this mock.
```

File: mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassinListener.java [hunk 9]
```diff
@@ -68,32 +81,49 @@ public ExecutionMode getExecutionMode() {
...
-                spamAssassin.learnHam(messages, event.getUser());
+            if (event instanceof Added) {
+                Added addedEvent = (Added) event;
+                if (isAppendedToInbox(addedEvent)) {
+                    Mailbox mailbox = mapperFactory.getMailboxMapper(session).findMailboxById(addedEvent.getMailboxId());
+                    MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
+
+                    List<InputStream> contents = addedEvent.getUids()
+                        .stream()
+                        .flatMap(uid -> retrieveMessage(messageMapper, mailbox, uid))
```
Comment (mbaechler) [inline mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassinListener.java]: why don't we group them by message ranges ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java [hunk 10]
```diff
@@ -1,162 +1,157 @@
...
+    public void event(Event event) {
+        try {
+            MailboxSession session = mailboxManager.createSystemSession(event.getUser().asString());
+            if (event instanceof MessageEvent) {
+                if (event instanceof EventFactory.AddedImpl) {
+                    EventFactory.AddedImpl added = (EventFactory.AddedImpl) event;
+                    Mailbox mailbox = added.getMailbox();
+
+                    for (MessageUid uid : added.getUids()) {
+                        retrieveMailboxMessage(session, mailbox, uid)
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java]: could we group them by message ranges ?

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**